package ammmo.gateway.protocol;

message AssociateDevice {
  required string device = 1;
  required string user = 2;
  required string key = 3;
}

message AssociateResult {
  enum Status {
    FAILED = 0;
    SUCCESS = 1;
  }
  required Status result = 1;
  optional string message = 2;
}

message PushData {
  required string uri = 1;
  optional string mime_type = 2;
  optional bytes data = 3;
}

message RegisterDataInterest {
  required string uri = 1;
}

message UnregisterDataInterest {
  required string uri = 1;
}

message GatewayWrapper {
  enum MessageType {
    ASSOCIATE_DEVICE = 0;
    ASSOCIATE_RESULT = 1;
    PUSH_DATA = 2;
    REGISTER_DATA_INTEREST = 3;
    UNREGISTER_DATA_INTEREST = 4;
  }
  
  required MessageType type = 1;
  
  optional AssociateDevice associate_device = 2;
  optional AssociateResult associate_result = 6;
  optional PushData push_data = 3;
  optional RegisterDataInterest register_data_interest = 4;
  optional UnregisterDataInterest unregister_data_interest = 5;
}
